README.md
# terraformscaffold

A framework for controlling multi-environment multi-component terraform-managed AWS infrastructure

Requires terrafrom version 0.9 or later and, if on a Mac, gnu-getopt (brew install gnu-getopt then add to path)

## Overview

Terraform scaffold consists of a terraform wrapper bash script, a bootstrap script and a set of directories providing the locations to store terraform code and variables.

| Component | Description |
|-------|------------------------|
| bin/terraform.sh | The terraformscaffold script |
| bootstrap/ | The bootstrap terraform code used for creating the terraformscaffold S3 bucket |
| components/ | The location for terraform "components". Terraform code intended to be run directly as a root module. |
| etc/ | The location for environment-specific terraform variables files:<br/>`env_{region}_{environment}.tfvars`<br/>`versions_{region}_{environment}.tfvars` |
| lib/ | Optional useful libraries, such as Jenkins pipeline groovy script |
| modules/ | The optional location for terraform modules called by components |
| src/ | The optional location for source files, e.g. source for lambda functions zipped up into artefacts inside components |

## Concepts & Assumptions

### Multi-Component Environment Concept

The Scaffold is built around the concept that a logical "environment" may consist of any number of independent components across independent AWS accounts. What provides consistency across these components, and therefore defines the environment, are the variables shared between the components. For example, the CIDR block defining a primary VPC for a production environment is needed by the component that creates the VPC, it may also be needed by components in other VPCs or accounts. All components in a production enviroment are likely to share the "envirornment" variable value "production".

Scaffold achieves this by maintaining variables specific to an environment in the file _etc/env_{region}_{environment}.tfvars_, and then providing those variables as inputs to all components run for that environment. Any variables not required by a component are safely ignored, but all components have visibility of all variables for an environment.

Important Note: Variables in the Env and Versions variables files are not merged. You cannot, for example, define the same map variable in both files and have the keys in each definition merged into a resulting set. When a variable is defined more than once, the last one terraform evaluates takes precedence and overrides any previously encountered definition. To prevent unexpected consequences, never define variable values in more than one place.

### State File Location Consistency and Referencing

Scaffold uses AWS S3 for storage of tfstate files. The bucket and location are specifically defined to be predictable to organise their storage and permit the use of terraform_remote_state resources. Any scaffold component can reliably refer to the S3 location of the state file for another component and depend upon the outputs provided. The naming convention is:     _s3://${backend_name}/${project}/${aws_region}/${environment}/${component}.tfstate_.

Each functional scaffold S3 bucket will only therefore contain content within the keyspace _/${project}/${aws_region}_ as these are unique to the bucket as well as all contents. The reason for the use of this keyspace is to permit the aggregation of state files from multiple bucket into a master bucket for backup or read-only review purposes. All scaffold buckets relevant to a person or organisation could be safely synchronised to a single bucket without fear of keyspace overlap.

### Variables Files: Environment & Versions

Scaffold provides a logical separation of several types of environment variable:
 * Global variables
 * Region-scoped global variables
 * Local variables (local to a component, common across environments)
 * Static environment variables
 * Frequently-changing versions variables
 * Dynamic (S3 stored) variables


This seperation is purely logical, not functional. It makes no functional difference in which file a variable lives, or even whether a versions variables file exists; but it provides the capacity to seperate out mostly static variables that define the construction of the environment from variables that could change on each apply providing new AMI IDs, or dockerised application versions or database snapshot IDs when recreating development and testing databases.

### AWS Credentials

Terraform Scaffold does not provide any mechanism for running terraform across multiple AWS accounts simultaneously, storing state files in a different account to the account being modified or any other functionality that would require Scaffold to intelligently manage AWS credentials. After extensive research and development it has become apparant that, despite some features available in terraform to handle more than one AWS account and the use of IAM roles, the features are not sufficiently mature or flexible to allow their application in a generic form.

Therefore, to ensure widest possible reach and capability of Scaffold, it requires that a specific set of AWS credentials be provided to it at invocation. These credentials must have the necessary access to read and write to the bootstrapped S3 state file bucket, and to create/modify/destroy the AWS resources being controlled via terraform.

Credentials can be provided in any of the standard mechanisms provided by the Boto search path, for example, EC2 Instance Profiles, an _~/.aws/credentials_ file or _AWS_ACCESS_KEY_ID_ and _AWS_SECRET_ACCESS_KEY_ environment variables.

If you want to make use of instance profiles, MFA tokens, AWS STS, Cross Account Roles or other fantastic IAM trickery, the recommended practice is to use a static access key or instance profile to call AWS STS using the AWS CLI tools, and then assign the temporary credentials that are generated to the AWS credential environment variables so that terraform can make use of them. This is done externally to Scaffold and would normally be integrated into a Jenkins job as a step to perform to prepare the environment before calling Scaffold.

### pre_apply.sh & post_apply.sh

Although as yet somewhat unrefined, Scaffold provides the capacity to incorporate additional scripted actions to take prior to and after running terraform on a given component. If there is a file called "pre_apply.sh" present in the top level of the component you are working with, then it will be executed as a bash script prior to any terraform action. If a file called post_apply.sh is present it will be executed immediately following any terraform action. This capability clearly could do with some improvement to support complex deployments with script dependencies, but as yet I have none to play with.


## Usage
### Bootstrapping
Before using Scaffold, a bootstrapping stage is required. Scaffold is responsible for creating and maintaining the S3 buckets it uses to store component state files and even keeps the state file that defines the scaffold bucket in the same bucket. This is done with a special bootstrap mode within the script, invoked with the '--bootstrap' parameter. When used with the "apply" action, this will cause the script to create a bootstrap bucket and then configure the bucket as a remote state location for itself. nd upload the tfstate used for managing the bucket to the bucket. Once created, the bucket can then be used for any terraform apply for the specific combination of project, region and AWS account.

It is not recommended to modify the bootstrap code after creation as it risks the integrity of the state files stored in the bucket that manage other deployments; however this can be mitigated by configuring synchronisation with a master backup bucket external to Scaffold management.

Bootstrapping usage:

```bash
bin/scaffold.sh \
    --bootstrap
    -p/--project `project` \
    -b/--bucket-name `bucket_name` \
    -r/--region `region` \
    -a/--action plan
```

```bash
bin/scaffold.sh \
    --bootstrap
    -p/--project `project` \
    -b/--bucket-name `bucket_name` \
    -r/--region `region` \
    -a/--action apply
```

Where:
* `project`: the name of the project to have a terraform bootstrap applied
* `bucket_name` (optional override for the bucket name): Defaults to: `${aws_account_alias}-scaffold if an alias is defined, else "${project}-${aws_account_id}-scaffold""`
* `region` (optional): Defaults to value of the AWS_DEFAULT_REGION environment variable

If the bucket should be encrypted with a custom CMK, provide they key as an additonal arg -- -var 'kms_key=[keyid]'

### Running

The terraformscaffold script is invoked as bin/terraform.sh. Once a state bucket has been bootstrapped, bin/terraform.sh can be run to apply terraform code. Its usage as of 25/01/2017 is:

```bash
bin/terraform.sh \
  -a/--action        `action` \
  -b/--bucket-name   `bucket_name` \
  -c/--component     `component_name` \
  -e/--environment   `environment` \
  -g/--group         `group` (optional) \
  -i/--build-id      `build_id` (optional) \
  -p/--project       `project` \
  -r/--region        `region` \
  -- \
  <additional arguments to forward to the terraform binary call>
```

Where:
* `action`: Terraform action (or pseudo-action) to take, e.g. plan, apply, plan-destroy (runs plan with the -destroy flag), destroy, show
* `bucket_name` (optional): Defaults to: `${aws_account_alias}-scaffold if an alias is defined, else "${project}-${aws_account_id}-scaffold""` - Only for use where a different bucket prefix has been bootstrapped
* `build_id` (optional): Used in conjunction with the plan and apply actions, `build_id` causes the creation and consumption of terraform plan files (.tfplan)
  * When `build_id` is omitted:
    * "Plan" provides normal plan output without generating a plan file
    * "Apply" directly applies the component based on the code and state it is given.
  * When `build_id` is provided:
    * "Plan" creates a plan file with `build_id` as part of the file name, and uploads the plan to the S3 state bucket under a key called "plans/" alongside the corresponding state file
    * "Apply" looks for and downloads the corresponding plan file generated by a plan job, and applies the changes in the plan file
  * It is usual to provide, for example, the Jenkins _$BUILD_ID_ parameter to Plan jobs, and then manually reference that particular Job ID when running a corresponding apply job.
* `component_name`: The name of the terraform component in the components directory to run the `action` against.
* `environment`: The name of the environment the component is to be actioned against, therefore implying the variables file(s) to be included
* `group` (optional): The name of the group to which the environment belongs, permitting the use of a group tfvars file as a "meta-environment" shared by more than one environment
* `project`: The name of the project being deployed, as per the default bucket-prefix and state file keyspace
* `region` (optional): The AWS region name unique to all components and terraform processes. Defaults to the value of the _AWS_DEFAULT_REGION_ environment variable.
* `additional arguments`: Any arguments provided after "--" will be passed directly to terraform as its own arguments, e.g. allowing the provision of a 'target=value' parameter.

CHANGELOG.md
## 1.4.3 (17/05/2018)

 * Remove account id from default bucket name as it unnecessarily increases the attack surface. Can manually add uniqueness with project name

## 1.4.2 (05/02/2018)

 * Add an init with -backend=false in bootstrap mode, otherwise providers are not downloaded
 * Fix bootstrap output that was using a now deprecated bucket resource attribute
 * Add an indicative .terraform-version file for tfenv support in bootstrap

## 1.4.1 (18/12/2017)

 * Remove `-upgrade` from init. Added prematurely. Can go back in when 0.9 support is dropped.

## 1.4.0 (18/12/2017)

 * Support terraform 0.10/0.11, bypassing new built-in approval mechanism.
 * Explicitly cache plugin downloads.
 * Set TF_IN_AUTOMATION.

## 1.3.1 (27/07/2017)

 * Change from error to warn on non-presence of requested group variables file

## 1.3.0 (26/07/2017)

 * Introduce the group variables file functionality

## 1.2.0 (09/06/2017)

 * Merge bootstrap functionality into the main script

## 1.1.4 (16/05/2017)

 * Global and Region scoped variables files

## 1.1.3 (12/04/2017)

 * Support the use of .terraform-version file in components when in the presence of [tfenv](https://github.com/kamatama41/tfenv)

## 1.1.2 (06/04/2017)

 * Provide a case for import that requires the variable file parameters

## 1.1.1 (06/04/2017)

 * Fix 0.9.2+ support for bootstrap.sh

## 1.1.0 (28/03/2017)

 * Move from terraform-0.8 to terraform-0.9
   * Change remote state from manual config to temporary-file "backend"
   * Complain in the code comments about Hashicorp forcing my hand on this
   * Don't push state any more; there's no persisted local copy to push
 * Same change for bootstrap.sh (UNTESTED!)

## 1.0.1 (15/03/2017)

 * Bugfix: Duplicate variable warning conditional to presence of duplicates

## 1.0.0 (14/03/2017)

 * Add CHANGELOG.md
 * Add version and help parameters to bin/terraform.sh
 * Add optional unencrypted S3 parameters
 * Move secrets S3 key path to more appropriate place
 * Test and Warn on duplicate variables
 * Support arbitrary terraform actions
 * Additional comments


components/bastion/template/sbx-user_data.sh
#!/bin/bash

# Install aws-ec2-ssh (https://github.com/widdix/aws-ec2-ssh)
wget -O /opt/authorized_keys_command.sh https://raw.githubusercontent.com/widdix/aws-ec2-ssh/master/authorized_keys_command.sh && \
	chmod +x /opt/authorized_keys_command.sh

wget -O /opt/import_users.sh https://raw.githubusercontent.com/widdix/aws-ec2-ssh/master/import_users.sh && \
	chmod +x /opt/import_users.sh

# Create aws-ec2-ssh config
cat <<EOF > /etc/aws-ec2-ssh.conf
IAM_AUTHORIZED_GROUPS="${environment}-${region}-iam-ssh-users"
LOCAL_MARKER_GROUP="iam-synced-users"
LOCAL_GROUPS=""
SUDOERS_GROUPS=""
ASSUMEROLE=""

# Remove or set to 0 if you are done with configuration
# To change the interval of the sync change the file
# /etc/cron.d/aws-ec2-ssh
DONOTSYNC=0
EOF

# Enable aws-ec2-ssh
sed -i 's:#AuthorizedKeysCommand none:AuthorizedKeysCommand /opt/authorized_keys_command.sh:g' /etc/ssh/sshd_config
sed -i 's:#AuthorizedKeysCommandUser nobody:AuthorizedKeysCommandUser nobody:g' /etc/ssh/sshd_config

# Create aws-ec2-ssh cron job to import users
echo "*/10 * * * * root /opt/import_users.sh" > /etc/cron.d/import_users && \
	chmod 0644 /etc/cron.d/import_users

service sshd restart


components/bastion/template/prod-user_data.sh
#!/bin/bash

# Create the index maintainance script
cat >/home/ec2-user/deleteindex.sh <<'EOT'
index=cwl-`date --date='1 week ago' '+%Y.%m.%d'`
echo $index
curl -XDELETE "https://prod-elasticsearch.sptlfe-prod.com/$index"
EOT

# Create the user service my stable mail trigger
cat >/home/ec2-user/triggermail.sh <<'EOT'
curl -X GET user-service.private:8080/my-stable/email-job
EOT

#Install the cron job for above
echo "0 22 * * * ec2-user /home/ec2-user/deleteindex.sh > /home/ec2-user/deleteindex.log" > /etc/cron.d/ec2-user_jobs
echo "0 17 * * * ec2-user /home/ec2-user/triggermail.sh >> /home/ec2-user/triggermail.log" >> /etc/cron.d/ec2-user_jobs

# Install aws-ec2-ssh (https://github.com/widdix/aws-ec2-ssh)
wget -O /opt/authorized_keys_command.sh https://raw.githubusercontent.com/widdix/aws-ec2-ssh/master/authorized_keys_command.sh && \
	chmod +x /opt/authorized_keys_command.sh

wget -O /opt/import_users.sh https://raw.githubusercontent.com/widdix/aws-ec2-ssh/master/import_users.sh && \
	chmod +x /opt/import_users.sh

# Create aws-ec2-ssh config
cat <<EOF > /etc/aws-ec2-ssh.conf
IAM_AUTHORIZED_GROUPS="${environment}-${region}-iam-ssh-users"
LOCAL_MARKER_GROUP="iam-synced-users"
LOCAL_GROUPS=""
SUDOERS_GROUPS=""
ASSUMEROLE=""

# Remove or set to 0 if you are done with configuration
# To change the interval of the sync change the file
# /etc/cron.d/aws-ec2-ssh
DONOTSYNC=0
EOF

# Enable aws-ec2-ssh
sed -i 's:#AuthorizedKeysCommand none:AuthorizedKeysCommand /opt/authorized_keys_command.sh:g' /etc/ssh/sshd_config
sed -i 's:#AuthorizedKeysCommandUser nobody:AuthorizedKeysCommandUser nobody:g' /etc/ssh/sshd_config

# Create aws-ec2-ssh cron job to import users
echo "*/10 * * * * root /opt/import_users.sh" > /etc/cron.d/import_users && \
	chmod 0644 /etc/cron.d/import_users

service sshd restart


components/techtest/template/dev-user_data.sh
#!/bin/bash

# Create the index maintainance script
cat >/home/ec2-user/deleteindex.sh <<'EOT'
index=cwl-`date --date='1 week ago' '+%Y.%m.%d'`
echo $index
curl -XDELETE "https://prod-elasticsearch.sptlfe-dev.com/$index"
EOT


# Add DNA team keys to ec2-user
cat >> /home/ec2-user/.ssh/authorized_keys <<EOL
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDdQnkd7T4YFAyZ622SGNBI8u/pdEdPtSr/LE5lNaaO21Xivp5GbCd40+oFWybIoeFyyx5jtsjI42lGEv5P/2wFZnMA+UtxpyjX57xfBwbGK5x4qrsU5XxwxPqbvT1CWbw8Psmjn1w7xSxDJ6iAOQMgb6K4fYq2XvNhKh8nL+eMGHELsK6MMjVfEW4noUrWRURowGqlecuSOCIgAzJZqYY0RmOADh7ZINRi6433llMfG+Nj727XOw09f8XzvIrXCq3Y5DTvZQ1RFvbbx49tYz/mAHBLL5IH+VRVPvaKieNqYzVhVVfjZA3/IZnQcmEMvXRPjTYGm+r+juZQGYw8LDPPxtWYaAoOXyietXce03DMhRN3K3iRsT8eWfRI437Xcmqrdkevgla2q1tjp0zFSoXItg5kopL2SDx1pjVlDzOGQiCO1DIZBHOIFiANoDRSf03gxGUvPo2/1+JqQly6dEfyUjp4zKyYZcWoqx4BPPlUZ4JUkIMWUiu8aP8zrBvcrXMgLBXE60eDCNlO26FIjzdX85NW6aeEim59DpgnMnpSdd8xbJ2D7KSvj1BRcWO47EjYeA+cx4ToG1V2N4Fmnh5//Q1ArQH9uKezDhnhrIlT5WEv0Q1F1S2/KHpgSNrBlGb5c2hwg88+ZLP3fnJ6Yay13VwSOhLP/8p0twQ25dh60w== antonio.quintana@nagra.com
EOL

components/bastion/template/mgmt-dev-user_data.sh
#!/bin/bash

# Install aws-ec2-ssh (https://github.com/widdix/aws-ec2-ssh)
wget -O /opt/authorized_keys_command.sh https://raw.githubusercontent.com/widdix/aws-ec2-ssh/master/authorized_keys_command.sh && \
	chmod +x /opt/authorized_keys_command.sh

wget -O /opt/import_users.sh https://raw.githubusercontent.com/widdix/aws-ec2-ssh/master/import_users.sh && \
	chmod +x /opt/import_users.sh

# Create aws-ec2-ssh config
cat <<EOF > /etc/aws-ec2-ssh.conf
IAM_AUTHORIZED_GROUPS="${environment}-${region}-iam-ssh-users"
LOCAL_MARKER_GROUP="iam-synced-users"
LOCAL_GROUPS=""
SUDOERS_GROUPS=""
ASSUMEROLE=""

# Remove or set to 0 if you are done with configuration
# To change the interval of the sync change the file
# /etc/cron.d/aws-ec2-ssh
DONOTSYNC=0
EOF

# Enable aws-ec2-ssh
sed -i 's:#AuthorizedKeysCommand none:AuthorizedKeysCommand /opt/authorized_keys_command.sh:g' /etc/ssh/sshd_config
sed -i 's:#AuthorizedKeysCommandUser nobody:AuthorizedKeysCommandUser nobody:g' /etc/ssh/sshd_config

# Create aws-ec2-ssh cron job to import users
cat "*/10 * * * * root /opt/import_users.sh" > /etc/cron.d/import_users && \
	chmod 0644 /etc/cron.d/import_users

service sshd restart


components/automation-jenkins/template/user_data.sh
#!/bin/bash

####################################################
#
# BEGIN STANDARD USERDATA
#
####################################################

exec > >(tee /var/log/user-data.log|logger -t user-data -s 2>/dev/console) 2>&1
echo BEGIN
date '+%Y-%m-%d %H:%M:%S'
echo END

# Update the MOTD

cat > /etc/motd <<EOL

####################################################

        IMMUTABLE SERVER, DO NOT PATCH WITH YUM

     1. Create new AMI with packer
     2. Update the versions file, 
     3. Apply new config with terraform.

####################################################

EOL

# Clean up all but the latest kernel
sudo package-cleanup -y --oldkernels --count=1


####################################################
#
# This Jenkins install uses EFS to decouple state for HA
#
####################################################

# Configure EFS
yum install -y amazon-efs-utils
mkdir /mnt/efs
mount -t efs -o tls ${efs_id}:/ /mnt/efs
mkdir /mnt/efs/jenkins

# Install Jenkins
yum update –y
yum remove -y java-1.7.0-openjdk
yum install -y java-1.8.0
yum install git

wget -O /etc/yum.repos.d/jenkins.repo http://pkg.jenkins-ci.org/redhat/jenkins.repo
rpm --import https://pkg.jenkins.io/redhat/jenkins.io.key
yum install jenkins -y
chown jenkins.jenkins /mnt/efs/jenkins
## Set Jenkins Home before we start Jenkins
sed -i 's/\/var\/lib\/jenkins/\/mnt\/efs\/jenkins/g' /etc/sysconfig/jenkins
chkconfig jenkins on
service jenkins start


components/bastion/template/uat-user_data.sh
#!/bin/bash

# Create the index maintainance script
cat >/home/ec2-user/deleteindex.sh <<'EOT'
index=cwl-`date --date='1 week ago' '+%Y.%m.%d'`
echo $index
curl -XDELETE "https://prod-elasticsearch.sptlfe-dev.com/$index"
EOT

# Create the user service my stable mail trigger
cat >/home/ec2-user/triggermail.sh <<'EOT'
curl -X GET user-service.private:8080/my-stable/email-job
EOT

#Install the cron job for above
#echo "0 22 * * * ec2-user /home/ec2-user/deleteindex.sh > /home/ec2-user/deleteindex.log" > /etc/cron.d/ec2-user_jobs
echo "0 17 * * * ec2-user /home/ec2-user/triggermail.sh >> /home/ec2-user/triggermail.log" >> /etc/cron.d/ec2-user_jobs

# Install aws-ec2-ssh (https://github.com/widdix/aws-ec2-ssh)
wget -O /opt/authorized_keys_command.sh https://raw.githubusercontent.com/widdix/aws-ec2-ssh/master/authorized_keys_command.sh && \
	chmod +x /opt/authorized_keys_command.sh

wget -O /opt/import_users.sh https://raw.githubusercontent.com/widdix/aws-ec2-ssh/master/import_users.sh && \
	chmod +x /opt/import_users.sh

# Create aws-ec2-ssh config
cat <<EOF > /etc/aws-ec2-ssh.conf
IAM_AUTHORIZED_GROUPS="${environment}-${region}-iam-ssh-users"
LOCAL_MARKER_GROUP="iam-synced-users"
LOCAL_GROUPS=""
SUDOERS_GROUPS=""
ASSUMEROLE=""

# Remove or set to 0 if you are done with configuration
# To change the interval of the sync change the file
# /etc/cron.d/aws-ec2-ssh
DONOTSYNC=0
EOF

# Enable aws-ec2-ssh
sed -i 's:#AuthorizedKeysCommand none:AuthorizedKeysCommand /opt/authorized_keys_command.sh:g' /etc/ssh/sshd_config
sed -i 's:#AuthorizedKeysCommandUser nobody:AuthorizedKeysCommandUser nobody:g' /etc/ssh/sshd_config

# Create aws-ec2-ssh cron job to import users
echo "*/10 * * * * root /opt/import_users.sh" > /etc/cron.d/import_users && \
	chmod 0644 /etc/cron.d/import_users

service sshd restart
